### PR TITLE
feat: Observe MLS settings change (WPB-308)

### PR DESCRIPTION
**Pull request by borichellow**
# What's new in this PR?

### Issues

- need an option to observe when MLS settings changed and inform user about it.
- need an option to snooze dialog about that changes
- need an option to set a time after which user couldn't snooze anymore

### Causes (Optional)

Was not implemented yet

### Solutions

Save MLS enabling into Preferences with a 2 fields:
1. time after which user should be notified about changes `notifyUserAfter`
2. time when user has to set up E2E identity `enablingDeadline`

Add UseCases for observing MLS settings changes and for snoozing this dialog (updating `notifyUserAfter` field)

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/MarkMLSEnableChangeAsNotifiedUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * Mark the MLS enabling status change as notified
+ * need to be called after notifying the user about the change
+ * e.g. after showing a dialog, or a toast etc.
+ */
+interface MarkMLSEnableChangeAsNotifiedUseCase {
+    suspend operator fun invoke()
+}
+
+class MarkMLSEnableChangeAsNotifiedUseCaseImpl(
```
Comment (yamilmedina) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/MarkMLSEnableChangeAsNotifiedUseCase.kt:31]:
```suggestion
internal class MarkMLSEnableChangeAsNotifiedUseCaseImpl(
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/MarkMLSEnableChangeAsNotifiedUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    private val userConfigRepository: UserConfigRepository
+) : MarkMLSEnableChangeAsNotifiedUseCase {
+
+    override suspend fun invoke() {
```
Comment (yamilmedina) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/MarkMLSEnableChangeAsNotifiedUseCase.kt:35]: Question, do we need to return something to know what to do after, or is it some mechanism, like fire and forget?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/ObserveMLSEnabledUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * Observe [MLSEnablingSetting] to notify user when setting is changed
+ */
+interface ObserveMLSEnabledUseCase {
+    /**
+     * @return [Flow] of [MLSEnablingSetting]
+     */
+    operator fun invoke(): Flow<MLSEnablingSetting>
+}
+
+class ObserveMLSEnabledUseCaseImpl(
```
Comment (yamilmedina) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/ObserveMLSEnabledUseCase.kt:37]:
```suggestion
internal class ObserveMLSEnabledUseCaseImpl(
```

**Review by yamilmedina — CHANGES_REQUESTED** [review]
Good job, just some questions and small suggestions 🙌

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/MarkMLSEnableChangeAsNotifiedUseCase.kt (hunk 2 above)
Comment (borichellow) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/MarkMLSEnableChangeAsNotifiedUseCase.kt:35]: yeah, i belive: shoot and forget is enough for now

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/event/Event.kt [hunk 4]
```diff
@@ -471,6 +472,19 @@ sealed class Event(open val id: String, open val transient: Boolean) {
             )
         }
 
+        data class MLSE2EIdUpdated(
+            override val id: String,
+            override val transient: Boolean,
+            val model: MLSE2EIdModel
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/event/Event.kt:478]: let's have the naming always without the d in Identity, I know the backend gives it like that, but we don't call it like that.
```suggestion
            val model: MLSE2EIModel
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/MarkMLSE2EIdEnableChangeAsNotifiedUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.logic.feature.user
+
+import com.wire.kalium.logic.configuration.UserConfigRepository
+import kotlin.time.DurationUnit
+import kotlin.time.toDuration
+
+/**
+ * Mark the MLS enabling status change as notified
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/MarkMLSE2EIdEnableChangeAsNotifiedUseCase.kt:25]: please update the comments to cover e2ei.

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/base/authenticated/featureConfigs/FeatureConfigResponse.kt [hunk 6]
```diff
@@ -51,7 +51,9 @@ data class FeatureConfigResponse(
     @SerialName("validateSAMLemails")
     val validateSAMLEmails: FeatureConfigData.ValidateSAMLEmails,
     @SerialName("mls")
-    val mls: FeatureConfigData.MLS?
+    val mls: FeatureConfigData.MLS?,
+    @SerialName("mlsE2EId")
+    val mlsE2EId: FeatureConfigData.MLSE2EId
```
Comment (mchenani) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/base/authenticated/featureConfigs/FeatureConfigResponse.kt:56]: here the same.

**Review by mchenani — CHANGES_REQUESTED** [review]
some namings, and one suggestion: please make sure this works on ApiV0 and ApiV4.

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/client/MarkMLSE2EIdEnableChangeAsNotifiedUseCaseTest.kt [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+import io.mockative.any
+import io.mockative.eq
+import io.mockative.given
+import io.mockative.mock
+import io.mockative.once
+import io.mockative.verify
+import kotlinx.coroutines.test.runTest
+import kotlin.test.Test
+
+class MarkMLSE2EIdEnableChangeAsNotifiedUseCaseTest {
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/client/MarkMLSE2EIdEnableChangeAsNotifiedUseCaseTest.kt:34]:
```suggestion
class MarkMLSE2EIEnableChangeAsNotifiedUseCaseTest {
```

TBH I can't understand the name. Mark MLS E2EI Enable (?) Change (?)

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/config/UserConfigStorage.kt [hunk 8]
```diff
@@ -86,6 +86,21 @@ interface UserConfigStorage {
...
+
+    /**
+     * get MLSE2EIdSetting
+     */
+    fun getMLSE2EIdSetting(): MLSE2EIdSettingEntity
+
+    /**
+     * get Flow of the saved MLSE2EIdSetting
+     */
+    fun mlsE2EIdSettingFlow(): Flow<MLSE2EIdSettingEntity>
```
Comment (vitorhugods) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/config/UserConfigStorage.kt:102]:
```suggestion
    /**
     * save MLSE2EIdSetting
     */
    fun setMLSE2EISetting(settingEntity: MLSE2EIdSettingEntity)

    /**
     * get MLSE2EIdSetting
     */
    fun getMLSE2EISetting(): MLSE2EIdSettingEntity

    /**
     * get Flow of the saved MLSE2EIdSetting
     */
    fun mlsE2EISettingFlow(): Flow<MLSE2EIdSettingEntity>
```

**Review by vitorhugods — CHANGES_REQUESTED** [review]

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/UserConfigRepository.kt [hunk 9]
```diff
@@ -45,6 +48,10 @@ interface UserConfigRepository {
     fun getClassifiedDomainsStatus(): Flow<Either<StorageFailure, ClassifiedDomainsStatus>>
     fun isMLSEnabled(): Either<StorageFailure, Boolean>
     fun setMLSEnabled(enabled: Boolean): Either<StorageFailure, Unit>
+    fun getMLSE2EIdSetting(): Either<StorageFailure, MLSE2EIdSetting>
+    fun observeIsMLSE2EIdSetting(): Flow<MLSE2EIdSetting>
+    fun setMLSE2EIdSetting(setting: MLSE2EIdSetting): Either<StorageFailure, Unit>
+    fun snoozeMLSE2EIdNotification(timeMs: Long): Either<StorageFailure, Unit>
```
[the remark this answers is not in the document]
Comment (borichellow) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/UserConfigRepository.kt:54]: good point, Duration is better than specifying "Ms", "Mins" etc. :)

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/MLSE2EISetting.kt [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.logic.configuration
+
+import com.wire.kalium.persistence.config.MLSE2EISettingEntity
+import kotlinx.datetime.Instant
+
+data class MLSE2EISetting(
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/MLSE2EISetting.kt:23]: I suggest to remove MLS from it, let's have it more simple like only E2EI.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/UserConfigRepository.kt [hunk 11]
```diff
@@ -45,6 +48,10 @@ interface UserConfigRepository {
     fun getClassifiedDomainsStatus(): Flow<Either<StorageFailure, ClassifiedDomainsStatus>>
     fun isMLSEnabled(): Either<StorageFailure, Boolean>
     fun setMLSEnabled(enabled: Boolean): Either<StorageFailure, Unit>
+    fun getMLSE2EISetting(): Either<StorageFailure, MLSE2EISetting>
+    fun observeIsMLSE2EISetting(): Flow<Either<StorageFailure, MLSE2EISetting>>
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/UserConfigRepository.kt:52]:
```suggestion
    fun observeE2EISettings()...
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/featureConfig/SyncFeatureConfigsUseCase.kt [hunk 12]
```diff
@@ -160,4 +165,16 @@ internal class SyncFeatureConfigsUseCaseImpl(
...
     }
+
+    private fun handleMLSE2EISetting(featureConfig: MLSE2EIModel) {
+        val enablingDeadlineMs = featureConfig.config.verificationExpirationNS.toDuration(DurationUnit.NANOSECONDS).inWholeMilliseconds
+        userConfigRepository.setMLSE2EISetting(
+            MLSE2EISetting(
+                isRequired = featureConfig.status == Status.ENABLED,
+                discoverUrl = featureConfig.config.discoverUrl,
+                notifyUserAfter = DateTimeUtil.currentInstant(),
+                enablingDeadline = Instant.fromEpochMilliseconds(enablingDeadlineMs)
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/featureConfig/SyncFeatureConfigsUseCase.kt:176]: we call this value as GracePeriod, I suggest to adopt this terminology. thanks

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/v0/authenticated/FeatureConfigApiV0.kt [hunk 13]
```diff
@@ -23,6 +23,7 @@ import com.wire.kalium.network.api.base.authenticated.featureConfigs.FeatureConf
 import com.wire.kalium.network.api.base.authenticated.featureConfigs.FeatureConfigResponse
 import com.wire.kalium.network.utils.NetworkResponse
 import com.wire.kalium.network.utils.wrapKaliumResponse
+import io.ktor.client.call.body
```
Comment (mchenani) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/v0/authenticated/FeatureConfigApiV0.kt:26]: unneeded changes? I think this whole file doesn't need to be changed.

**Review by mchenani — CHANGES_REQUESTED** [review]
Looks good, Some naming suggestions.

**Review by vitorhugods — APPROVED** [review]
👨🏼‍🍳🤌🏻

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/MarkEnablingOfMLSE2EIAsNotifiedUseCase.kt [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+import com.wire.kalium.logic.configuration.UserConfigRepository
+import kotlin.time.Duration.Companion.days
+
+/**
+ * Mark the MLS End-to-End Identity enabling status change as notified
+ * need to be called after notifying the user about the change
+ * e.g. after showing a dialog, or a toast etc.
+ */
+interface MarkEnablingOfMLSE2EIAsNotifiedUseCase {
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/MarkEnablingOfMLSE2EIAsNotifiedUseCase.kt:28]:
```suggestion
interface MarkEnablingE2EIAsNotifiedUseCase {
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/MarkEnablingOfMLSE2EIAsNotifiedUseCase.kt [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    suspend operator fun invoke()
+}
+
+internal class MarkEnablingOfMLSE2EIAsNotifiedUseCaseImpl(
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/MarkEnablingOfMLSE2EIAsNotifiedUseCase.kt:32]:
```suggestion
internal class MarkEnablingE2EIAsNotifiedUseCaseImpl(
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/featureConfig/SyncFeatureConfigsUseCase.kt [hunk 16]
```diff
@@ -160,4 +165,16 @@ internal class SyncFeatureConfigsUseCaseImpl(
             )
         }
     }
+
+    private fun handleMLSE2EISettings(featureConfig: MLSE2EIModel) {
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/featureConfig/SyncFeatureConfigsUseCase.kt:169]:
```suggestion
    private fun handlE2EISettings(featureConfig: MLSE2EIModel) {
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/featureConfig/SyncFeatureConfigsUseCaseTest.kt [hunk 17]
```diff
@@ -570,6 +573,37 @@ class SyncFeatureConfigsUseCaseTest {
             })
     }
 
+    @Test
+    fun givenMlsE2EIIsDisasbled_whenSyncing_thenItShouldBeStoredAsDisabled() = runTest {
+        val mlsE2EIModel = MLSE2EIModel(MLSE2EIConfigModel("url", 10_000_000_000L), Status.DISABLED)
```
Comment (mchenani) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/featureConfig/SyncFeatureConfigsUseCaseTest.kt:578]:
```suggestion
        val mlsE2EIModel = E2EIModel(E2EIConfigModel("url", 10_000_000_000L), Status.DISABLED)
```

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/config/UserConfigStorage.kt [hunk 18]
```diff
@@ -168,6 +191,9 @@ class UserConfigStorageImpl(
     private val isGuestRoomLinkEnabledFlow =
         MutableSharedFlow<Unit>(extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)
 
+    private val mlsE2EIFlow =
```
Comment (mchenani) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/config/UserConfigStorage.kt:194]:
```suggestion
    private val e2EIFlow =
```

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/config/UserConfigStorage.kt [hunk 19]
```diff
@@ -239,6 +265,25 @@ class UserConfigStorageImpl(
 
     override fun isMLSEnabled(): Boolean = kaliumPreferences.getBoolean(ENABLE_MLS, false)
 
+    override fun setE2EISettings(settingEntity: E2EISettingsEntity) {
+        kaliumPreferences.putSerializable(
+            MLS_E2E_ID,
```
Comment (mchenani) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/config/UserConfigStorage.kt:270]:
```suggestion
            E2EI_SETTINGS,
```

**Review by mchenani — APPROVED** [review]
Awesome 🙇‍♂️ Just some final naming suggestions.

**Review by yamilmedina — APPROVED** [review]

**Pull request merged**